File: this-is-coding-test/18장_여행계획.js
class DisjointSet {
  constructor(n) {
    this.parent = new Array(n).fill(0);
    this.makeSet(n);
  }

  makeSet(n) {
    for (let i = 0; i < n; i++) {
      this.parent[i] = i
    }
  }

  findSet(x) {
    if (this.parent[x] !== x) {
      this.parent[x] = this.findSet(this.parent[x]);
    }

    return this.parent[x];
  }

  unionSet(x, y) {
    const px = this.findSet(x);
    const py = this.findSet(y);

    this.parent[py] = px;
  }
}

const getAnswer = () => {
  for (let i = 0; i < M - 1; i++) {
    if (disjointSet.findSet(i) !== disjointSet.findSet(i + 1)) {
      return 'NO';
    }
  }

  return 'YES';
}

const input = [
  '5 4',
  '0 1 0 1 1',
  '1 0 1 1 0',
  '0 1 0 0 0',
  '1 1 0 0 0',
  '1 0 0 0 0',
  '2 3 4 3',
];
// => YES

// const input = [
//   '6 4',
//   '0 1 1 0 0 0',
//   '1 0 0 0 0 0',
//   '1 0 0 0 0 0',
//   '0 0 0 0 1 0',
//   '0 0 0 1 0 1',
//   '0 0 0 0 1 0',
//   '3 4',
// ];
// => NO

const [N, M] = input.shift().split(' ').map(v => parseInt(v));
const roads = [];

for (let i = 0; i < N; i++) {
  roads.push(input.shift().split(' ').map(v => parseInt(v)));
}

// 0 ~ N - 1 로 전체 범위를 잡으면서
// 여행 계획의 숫자를 정수로 변환한 뒤 1을 빼준다
const plans = input.shift().split(' ').map(v => parseInt(v) - 1);

const roadsAdj = {};

for (let i = 0; i < N; i++) {
  roadsAdj[i] = [];
}

const disjointSet = new DisjointSet(N);

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    if (roads[i][j]) {
      // i, j 와 j, i 는 양방향으로 연결된다
      // 중복해서 unionSet 을 수행하지 않도록
      // i, j 의 부모 노드가 같으면
      // unionSet 을 수행하지 않는다
      if (disjointSet.findSet(i) !== disjointSet.findSet(j)) {
        disjointSet.unionSet(i, j);
      }
    }
  }
}

console.log(getAnswer());

File: this-is-coding-test/18장_여행계획.py
class DisjointSet:
    def __init__(self, n):
        self.p = [0] * n
        for k in range(n):
            self.make_set(k)

    def make_set(self, x):
        self.p[x] = x

    def find_set(self, x):
        if self.p[x] == x:
            return self.p[x]

        self.p[x] = self.find_set(self.p[x])
        return self.p[x]

    def union_set(self, x, y):
        px = self.find_set(x)
        py = self.find_set(y)

        self.p[py] = px


N, M = map(int, input().split())
disjointSet = DisjointSet(N)
travel_places = list()
for i in range(N):
    travel_place = list(map(int, input().split()))
    for j in range(N):
        if travel_place[j] == 1:
            disjointSet.union_set(i, j)

travel_plan = list(map(int, input().split()))
is_travel_possible = True
for i in range(1, M):
    x, y = travel_plan[i - 1] - 1, travel_plan[i] - 1
    px, py = disjointSet.find_set(x), disjointSet.find_set(y)
    if px != py:
        is_travel_possible = False
        break

if is_travel_possible:
    print('YES')
else:
    print('NO')
